example/src/UnionFind.hs

```diff
@@ -175,23 +175,26 @@ smm = StateMachineModel preconditions postconditions transitions initModel
 -- The generation of actions is parametrised by the number of @New@'s
 -- that have been generated.
 
-gen :: StateT Int Gen (Untyped Action (RefPlaceholder Void))
-gen = do
-  n <- get
-  if n == 0
-  then do
-    modify succ
-    lift $ Untyped . New <$> arbitrary
-  else do
-    cmd <- lift $ frequency
-      [ (1, Untyped . New  <$> arbitrary)
-      , (5, Untyped . Find <$> choose (0, n - 1))
-      , (5, Untyped <$> (Union <$> choose (0, n - 1) <*> choose (0, n - 1)))
-      ]
-    case cmd of
-      Untyped (New _) -> modify succ
-      _               -> return ()
-    return cmd
+gen :: Generator Void Action Int
+gen = Generator
+  { generator     = \n -> frequency
+     [ (1, Untyped .    New   <$> arbitrary)
+     , (5, Untyped .    Find  <$> choose (0, n - 1))
+     , (5, Untyped <$> (Union <$> choose (0, n - 1) <*> choose (0, n - 1)))
+     ]
+  , gprecondition = gprecondition'
+  , gtransition   = gtransition'
+  , initGenState  = 0
+  }
+  where
+  gprecondition' :: Int -> Action refs resp -> Bool
+  gprecondition' _ (New   _)   = True
+  gprecondition' n (Find  _)   = n > 0
+  gprecondition' n (Union _ _) = n > 0
```

Comment:
Possibily, that's already taken care of by the `StateMachineModel`'s `precondition` though... We probably want to merge the two concepts somehow...